openmmtools/states.py

```diff
@@ -2027,8 +2063,41 @@ def _read_context_state(self, context_state, check_consistency):
         self.box_vectors = openmm_state.getPeriodicBoxVectors(asNumpy=True)
         # Potential energy and kinetic energy must be updated
         # after positions and velocities or they'll be reset.
-        self.potential_energy = openmm_state.getPotentialEnergy()
-        self.kinetic_energy = openmm_state.getKineticEnergy()
+        self._potential_energy = openmm_state.getPotentialEnergy()
+        self._kinetic_energy = openmm_state.getKineticEnergy()
+        self._set_collective_variables(context_state)
+
+    def _set_collective_variables(self, context_state):
+        """
+        Update the collective variables from the context object
+
+        Parameters
+        ----------
+        context_state : simtk.openmm.Context or simtk.openmm.State
+            The object to read. This only works with Context's for now,
+            but in the future, this may support OpenMM State objects as well.
+        """
+        if isinstance(context_state, openmm.State):
```

Comment:
I think we should raise an exception if the user passes an `openmm.State` object and `get_collective_variables=True`. Otherwise, this will fail silently and end up with a value for `SamplerState.collective_variables` indistinguishable from having a `System` with no collective variables.

Reply:
I don't really like this idea. This can lead the user to be mandated to use the `get_collective_variables=False` in every call to this function when the collective variables are less common. 

> SamplerState.collective_variables indistinguishable from having a System with no collective variables.

Yes, though I thought that was partially the point? If a user never supplies the information about a given state, then we can't use that. I don't like the idea of the users being required to now track collective variables, which are an optional construct in of themselves.

Reply:
The user wouldn't have to set `get_collective_variable=False` in every call as long as he pass an actual `Context` object. Passing the `State` object is there exclusively for optimization (i.e. if I already have one around I don't have to read the info off the `Context` again).

In any case, I think avoiding silent failing would be worth forcing the users to add the extra option, and the default behavior should be the safest. With this implementation, for example, all our metropolized `MCMCMove`s are currently failing silently and not retrieving CV variables that should actually be available after `apply` has been performed.